lib/rouge/lexers/prolog.rb

```diff
@@ -12,7 +12,7 @@ class Prolog < RegexLexer
 
       state :basic do
         rule /\s+/, Text
-        rule /^#.*/, Comment::Single
+        rule /^#!.*/, Comment::Single
```

Comment:
Perhaps it would be more correct for it to be:

```suggestion
        rule /\A#!.*/, Comment::Single
```

 It seems as if a hash-prefaced line is treated as a comment if (and only if) that line is the first line in the file (see [here](http://www.swi-prolog.org/pldoc/man?section=plscript) and [here](http://gprolog.org/manual/gprolog.html#sec11)).

Reply:
Great catch. I would do this, but I don't think `\A` works the way we want due to how Rouge generates tokens (matching each rule against the remaining input). When I try it, non-first-line `#!`s are still matched, just like with `^`.

Reply:
Hmmm, you're right (and I think your logic is correct as to why).

I'm not able to experiment with anything at the moment but my immediate response is that the lexer should start in some kind of initial state which is only used for the first line and then after that, it goes into the normal beginning state. 

Or then again, maybe it's not really a big deal if people can generate comment tokens for things that aren't legitimate comments and I'm overthinking it :P

Reply:
Yep, `StringScanner` is not able to recognize either `^` or `\A` - it's a bug in Ruby that I haven't been able to get any traction on. We use the approach described above for other lexers, like `shell`.

Reply:
If things only are valid at the beginning of a line, we use a `:bol` state, like for `#preproc` statements in C/C++